ports/servoshell/app.rs

```diff
@@ -146,17 +165,26 @@ impl App {
                     app.servo = Some(servo);
                 }
 
-                // TODO does windows still need this workaround?
-                // https://github.com/emilk/egui/blob/9478e50d012c5138551c38cbee16b07bc1fcf283/crates/egui_glow/examples/pure_glow.rs#L203
-                // winit::event::Event::RedrawEventsCleared => todo!(),
                 winit::event::Event::RedrawRequested(_) => {
+                    // We need to redraw the window for some reason.
+                    trace!("RedrawRequested");
+
+                    // WARNING: do not defer painting or presenting to some later tick of the event
+                    // loop or servoshell may become unresponsive! (servo#30312)
+                    if need_recomposite {
+                        trace!("need_recomposite");
+                        app.servo.as_mut().unwrap().recomposite();
+                    }
                     if let Some(mut minibrowser) = app.minibrowser() {
                         minibrowser.update(window.winit_window().unwrap(), "RedrawRequested");
-
-                        // Tell Servo to repaint, which will in turn allow us to repaint the
-                        // minibrowser and present a complete frame without partial updates.
-                        app.event_queue.borrow_mut().push(EmbedderEvent::Refresh);
+                        minibrowser.paint(window.winit_window().unwrap());
+                    }
+                    if external_present {
+                        app.servo.as_mut().unwrap().present();
```

Comment:
Can we remove the `if external_present` condition, since that was just `let external_present = app.minibrowser.is_some();`?